src/util/Spotify.js

```diff
@@ -48,53 +48,66 @@ export const Spotify = {
         Tracklist.js, cannot read property 'map' of undefined, and pointing on the
         traks of the TrackList.. I couldn't understand why.*/
     async search (term) {
-      /* let userAccessToken = this.getUserAccessToken();   I commented this out becaues it is already defined in the global of this file*/
+      
+      const topic = 'Searching Spotify\'s API.'
+
+      const mainTerm = encodeURI(term); // this converts the term passed as a parameter to a URI format
+
+      let response;
+      let JSONResponse;
+      let error;
+      let mappedTracks;
 
       if (!userAccessToken){
         console.log('No Access Token.');
-        return;
+        return ;
       };
-
-      const mainTerm = encodeURI(term); // this converts the term passed as a parameter to a URI format
-      
       
+      //The code below tries to contact Spotifies API
       try {
-        const response =  await fetch(`https://api.spotify.com/v1/search?type=track&q=${mainTerm}`, {
+        response =  await fetch(`https://api.spotify.com/v1/search?type=track&q=${mainTerm}`, {
         headers: {Authorization: `Bearer ${userAccessToken}`}
-      })
-      
-      const jsonifyResponse = async (response) => {
-
-        let responseOK = await response.ok;
+      })      
+      } catch (e) {
+        error = new Error(`Network error, while ${topic}.`);
+        return Promise.reject(error);
+      }
 
-          if(responseOK){
-              return response.json()
-          }
-          throw new Error('Request failed');
+      if (!response.ok) {
+        error = new Error(`Server error, status: ${response.status}, while ${topic}.`);
+        return Promise.reject(error);
       }
-      const jsonResponse = await jsonifyResponse(response);
 
-      let responseTracks = await jsonResponse.tracks;
-      
-      if (!responseTracks){
-        return [];
+      // The code below will try to convert the response to JSON format
+
+      try{
+        JSONResponse = await response.json();
+
+      } catch (err){
+        error = new Error(`Fail to read response, while ${topic}.`);
+        return Promise.reject(error);
       }
 
+      
+      let searchResults =  () => {
 
-      return jsonResponse.tracks.items.map(track => ({
+        if (!JSONResponse.tracks){
+          mappedTracks =  [];
+        }
+        mappedTracks = JSONResponse.tracks.items.map(track => ({
+
+          id: track.id,
+          name: track.name,
+          artist: track.artists[0].name,
+          album: track.album.name,
+          uri: track.uri 
+      }));
+      };
 
-            id: track.id,
-            name: track.name,
-            artist: track.artists[0].name,
-            album: track.album.name,
-            uri: track.uri 
-        }));
-    } 
+      searchResults();

```

Comment:
instead of creating `mappedTracks` above, later creating a function that assigns it value, and calling it.

which is a 3 step process 

you can create `mappedTracks` right away :)

```
let mappedTracks = (JSONResponse.tracks && JSONResponse.tracks.items)
  ? (
  JSONResponse.tracks.items.map(track => ({
          id: track.id,
          name: track.name,
          artist: track.artists[0].name,
          album: track.album.name,
          uri: track.uri 
      }))
  )
  : []

```

Reply:
Cool, but I didn't understand why the condition checks both for JSONResponse.tracks and JSONResponse.tracks.items ? since the letter is nested inside the formar shouldn't it be enough to check only that?